Wire remaining tokens widget

chinins opened the pull request:
## Description

This PR wires remaining tokens widget.

Please, note that if you test it with real numbers and you have already purchased some tokens, the price status will always show either up or sold out. That's because the `target` tokens is negligibly small. I assume it's coming from the default parameters of the extension setup (as per my messages on discord). To be clarified with Raul when he is back. 

**New stuff** ✨

- Add `CurrentPeriodTokens` query, types & resolver
- - Add `CurrentPeriodTokens` query call to `CoinMachine`

**Changes** 🏗

- update `RemainingDisplayWidget` for the tokens data

## TODO

- Potentially refactor the `RemainingDisplayWidget` component once Armando merges his remaining time part. I think that there is potential for code improvement.

Resolves DEV-466

<img width="237" alt="Screenshot 2021-08-05 at 15 03 20" src="https://user-images.githubusercontent.com/34057551/128346759-a3eb1fba-ec8c-4d6c-bfef-26947f20b042.png">

alicjakujawa reviewed (changes requested):
Few comments here. I also created issue for updating init params, because i agree with you that they doesn't make sense currently.

ArmandoGraterol reviewed (commented):
I noticed while testing and working with my PR on top of this branch that there isn't a way to display the placeholder value for when the coin machine is empty/the sale hasn't started (which would be a plain old `0`).

chinins commented on src/modules/dashboard/components/CoinMachine/RemainingDisplayWidget/RemainingDisplayWidget.tsx:
```diff
@@ -95,12 +106,64 @@ const RemainingDisplayWidget = ({
+      return `${getFormattedTokenValue(
+        soldPeriodTokens,
+        decimals,
+      )}/${getFormattedTokenValue(maxPeriodTokens, decimals)}`;
```
as per my message on Discord, Karol confirmed it should be tokensSold/maxPerPeriod

chinins replied:
As per the discussion in product private, I've changed it

ArmandoGraterol commented on src/modules/dashboard/sagas/coinMachine/buyTokens.ts:
```diff
@@ -167,8 +174,18 @@ function* buyTokens({
       meta,
     });
   } catch (caughtError) {
+    yield;
```
What does this `yield` empty do here?

chinins replied:
good catch - removed

ArmandoGraterol reviewed (commented):
Re-tested it and now the `Tokens remaining` always shows `100/100` even though I bought 50 tokens :( 

Can any of you guys replicate it or is it just me?

alicjakujawa commented on src/data/resolvers/coinMachine.ts:
```diff
@@ -99,5 +99,40 @@ export const coinMachineResolvers = ({
+        return {
+          maxPerPeriodTokens: maxPerPeriodTokens.toString(),
+          activeSoldTokens:
+            activePeriod === currentPeriod ? activeSoldTokens.toString() : '',
```
Its not enough, you need to check if there was any TokensBought event in current period, if so then activeSoldTokens if not, then zero

alicjakujawa reviewed (changes requested):
yes i have same problem like @ArmandoGraterol  its most probably because of wrong logic i pointed out - see comment

chinins commented:
> Re-tested it and now the `Tokens remaining` always shows `100/100` even though I bought 50 tokens :(
> 
> Can any of you guys replicate it or is it just me?

You need to manually change the `isSale` variable in the `CoinMachine` period to `true` and then it would work as expected. The `0` shows only when there is no sale. The isSale logic will be implemented in a separate PR you as you probably know.

chinins commented:
Ready for review.

I have now added the automatic refresh of the period tokens when the period is over.

I think I also found a bug in the time remaining part. For me it never changed the font color to red. That was because the `value` is never updated unless we refresh the page. I fixed it by making it dependent on timeLeft and it's working properly now.

![period-update](https://user-images.githubusercontent.com/34057551/129266220-c2456af7-043d-4fd3-a75b-01a8a62fe477.gif)

ArmandoGraterol commented on src/modules/dashboard/components/CoinMachine/RemainingDisplayWidget/RemainingDisplayWidget.tsx:
```diff
@@ -66,21 +78,29 @@ const MSG = defineMessages({
     defaultMessage: 'Come back in...',
   },
+  soldOut: {
+    id: 'dashboard.CoinMachine.RemainingDisplayWidget.tokensTypeFooterText',
```
The id here should have `.soldOut` instead of `.tokensTypeFooterText`, right?

chinins replied:
good catch!

ArmandoGraterol commented on src/modules/dashboard/components/CoinMachine/RemainingDisplayWidget/RemainingDisplayWidget.tsx:
```diff
@@ -105,17 +125,76 @@ const RemainingDisplayWidget = ({
     typeof value === 'number' &&
     periodLength &&
-    ((value / 1000) * 100) / periodLength <= 10;
+    (timeLeft * 100) / periodLength <= 10;
```
Solid catch :100:

ArmandoGraterol reviewed (approved):
It's working nicely now! Looks good to me :+1:

alicjakujawa reviewed (commented):
Wanted to test it, but have this error in console, can you retest if you have this as well?

<img width="560" alt="Screenshot 2021-08-13 at 10 39 31" src="https://user-images.githubusercontent.com/13069555/129329752-a4a098be-286d-4e15-b02a-ee1e6825fb26.png">

chinins commented:
> Wanted to test it, but have this error in console, can you retest if you have this as well?
> 
> <img alt="Screenshot 2021-08-13 at 10 39 31" width="560" src="https://user-images.githubusercontent.com/13069555/129329752-a4a098be-286d-4e15-b02a-ee1e6825fb26.png">

I saw this error only this morning so I was convinced that it was from the feature/coin-machine-redux branch rebase (and I could somehow replicate it there when I tested as well). But many hours and lots of degrees of desperation later I managed to find and fix it (as always something small and not difficult).

You can review now.

alicjakujawa reviewed (changes requested):
Almost there, functionality seems to be working fine, i have only one comment before approving this

chinins commented on src/modules/dashboard/components/CoinMachine/RemainingDisplayWidget/RemainingDisplayWidget.tsx:
```diff
@@ -105,17 +125,76 @@ const RemainingDisplayWidget = ({
+      dispatch({
+        type: ActionTypes.COIN_MACHINE_PERIOD_UPDATE,
+        payload: { colonyAddress },
+      });
```
I think that we need it as the Tokens & Time parts should be separated. They just create messy code when they are together. That will be part of my refactor that I was talking about on discord.

alicjakujawa replied:
so if you will have addiitonal query, then you will use refetch from additional query, not additional saga for updating

chinins replied:
You can't do that, because the query needed won't be in the component and I don't think it's good to pass it via a prop to components (you said in your own comment to the other PR that it's good to avoid prop drilling). It's the same principle as in motions: time changes in one component & the query needs to be updated in a different component. You don't have this query now in that component anyway, it's in the coinMachine.

alicjakujawa replied:
Ok so maybe im not sure what you plan to do, but i will see in the PR itself. For now please use refetch provided by use query hook

chinins replied:
I am stating again that the query is not even in the component. It's in CoinMachine. Do you want me to pass it as a prop? I disagree with this approach.

rdig replied:
I actually agree with @chinins 's approach here. While I'm not a fan of calling sagas from components, in this case make sense.

As for just having this one saga just to update the time it makes it much better, as the code won't need to update values it won't use, so the update _should_ happen faster.

On the other hand, if we put more and more queries in this saga _(down the line)_, we might need to think about combining it with others, but for now I think it's OK if it stays like this.

However, I would make the time comparison something like:
```js
if (timeLeft <= 0 && colonyAddress) { ... }
```
This is just to guard against `timeLeft` underflowing below `0` _(something like `-1` or `-2`)_ in which case the logic check will break

alicjakujawa replied:
You can use this query in component itself, instead of passing props to the child. It will fetch from cached data so it wont cost anything

chinins replied:
That won't work as I need this query in `CoinMachine` for another CoinMachine child component.

I also think that Raul agreed with my approach so could we move on with this PR if you don't have any other objections?

alicjakujawa replied:
i mean you can use this query whenever you want, so what is the problem? We dont need additional resolvers and query in our codebase, we should use existing approaches.

chinins replied:
Yeah, but then I need it to be updated when the timer runs out and I will need this anyway. Or I can use it in CoinMachine once and then pass 1 value as prop to 2-3 components that need that. And use this refetching anyway as well. In any case I need this refetching mechanism.

We are wasting time here, it's become a too long and unproductive discussion. Can we already proceed? There are PRs that wait for it to be merged.

alicjakujawa replied:
We dont need additional code in our codebase. https://github.com/JoinColony/colonyDapp/blob/60026cf815138dc721e7115953d96abc0c16f4f1/src/modules/dashboard/components/ActionsPage/CountDownTimer/CountDownTimer.tsx#L179 this mechanism is exactly provided for cases like yours. And you said lastly that its important to have good codebase.

chinins replied:
I understand the method and I agree that it's important to have a good code base but I don't think that refetch should be used here as the refetch should be triggered in a component(s) different from where the query is. Unless we pass it as a prop which I don't think is a good idea.
